### ISSUE TITLE
Roslyn incorrectly reports incorrect warnings after finally (caused by merging and reporting unreachable state)

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51006]
**Version Used**: 
```
Branch master (30 Jan 2021)
Latest commit d0d9a59 by msftbot[bot]:
Merge pull request #50907 from dotnet/merges/release/dev16.9-to-master

Merge release/dev16.9 to master
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  string M()
  {
    string? s1 = "";
    try
    {
      return s1;
    }
    finally
    {
      s1 = null;
    }

    // CS0162: Unreachable code detected
    string s2 = s1; // CS8600: Converting null literal or possible null value to non-nullable type.
    if ("".Length > 0) return s2; // CS8603: Possible null reference return.
  }
}
```

**Expected Behavior**:
No nullability warnings.
The code never reaches the statements after finally because the method either returns or throws an exception.
Note that the compiler understands this invariant and reports CS0162 unreachable code warning and doesn't require return statements after `finally`

**Actual Behavior**:
`CS8600: Converting null literal or possible null value to non-nullable type.` and `CS8603: Possible null reference return.` warnings are reported from unreachable code.

**Notes**
The example above is minified to clearly demonstrate the problem. Here's a scenario similar to a real world one where the compiler's behavior leads to incorrect warnings.
```cs
using System;

#nullable enable
class C
{
  string? field;

  void M(bool condition)
  {
    if (field == null)
      return;

    if (condition)
    {
      try
      {
        // do something
        Console.WriteLine("exit");
        return;
      }
      finally
      {
        field = null; // reset state
      }
    }
    
    Console.WriteLine($"field value length = {field.Length}"); // false CS8602 Dereference of a possibly null reference
  }
}
```

The method checks that `field` is non-nullable, does something if some `condition` is satisfied and resets the `field` state e.g. using it as a flag that some activity has finished.
Note that the nullable state of `field` can never be observed after the `finally` block since it contains a `return` statement.
The code after the `if (condition)` statement should be able to dereference `field` without any problems since it's already checked above. However Roslyn incorrectly merges the nullable state of `field` on unreachable exit from the `finally` block to the nullability state after `if (condition)` producing incorrect warnings.

**Comment by jaredpar** [comment]
I'm not sure we've completely defined how NRT should work in unreachable code. Generally in unreachable code we've taken approaches to minimize extra warnings like assuming all values are definitely assigned. Should sit down and decide how we want this to behave here.

**Comment by RikkiGibson** [comment]
NullableWalker does define an unreachable state, where IIRC all variables have not-null state.

**Comment by TessenR** [comment]
I think the following two methods should have the same warnings. Looks like `M1` is already handling the unreachable code correctly. So my guess is that the state after finally with unreachable try block endpoint is currently not unreachable but should be.
```cs
#nullable enable
class C
{
  string M1()
  {
    string s1 = "";
    return s1;
      
    // CS0162: Unreachable code detected
    s1 = null; // no warnings
    
    string s2 = s1; // no warnings
    if ("".Length > 0) return s2; // no warnings
  }
    
  string M2()
  {
    string s1 = "";
    try
    {
      return s1;
    }
    finally
    {
    }

    // CS0162: Unreachable code detected
    s1 = null; // CS8600
    
    string s2 = s1; // CS8600
    if ("".Length > 0) return s2; // CS8603
  }
}
```
